File: tasks/translation/output/Qwen/Qwen2.5-Coder-32B-Instruct-AWQ/codenet/direct/multiline/comment/direct/Java/Python/s316260310_commented.py
import sys
import math

class Solver:
    def __init__(self):
        self.hp = Helper()
        self.MAXN = 1000006
        self.MOD = 10**9 + 7

    def solve(self):
        tc = self.hp.nextInt() if self.TESTCASES else 1
        for tce in range(1, tc + 1):
            self.solve_case(tce)

    def solve_case(self, tc):
        N = self.hp.nextInt()
        K = self.hp.nextLong()
        A = self.hp.getIntArray(N)

        straight = []
        cycle = []
        vis = [False] * N
        curr = 0
        while not vis[curr]:
            vis[curr] = True
            curr = A[curr] - 1

        st = 0
        while st != curr:
            straight.append(st)
            st = A[st] - 1

        cycle.append(curr)
        curr = A[curr] - 1
        while curr != cycle[0]:
            cycle.append(curr)
            curr = A[curr] - 1

        if K < len(straight):
            self.hp.println(straight[int(K)] + 1)
        else:
            K -= len(straight)
            self.hp.println(cycle[int(K % len(cycle))] + 1)

    TESTCASES = False


class Helper:
    def __init__(self):
        self.MOD = 10**9 + 7
        self.MAXN = 1000006
        self.rnd = random.Random()

    def getLongArray(self, size):
        return [self.nextLong() for _ in range(size)]

    def getIntArray(self, size):
        return [self.nextInt() for _ in range(size)]

    def getStringArray(self, size):
        return [self.next() for _ in range(size)]

    def joinElements(self, *ar):
        return ' '.join(map(str, ar))

    def gcd(self, a, b):
        return a if b == 0 else self.gcd(b, a % b)

    def max(self, *ar):
        return max(ar)

    def min(self, *ar):
        return min(ar)

    def sum(self, *ar):
        return sum(ar)

    def shuffle(self, ar):
        random.shuffle(ar)

    def reverse(self, ar):
        ar.reverse()

    def pow(self, base, exp, MOD):
        base %= MOD
        ret = 1
        while exp > 0:
            if exp & 1:
                ret = ret * base % MOD
            base = base * base % MOD
            exp >>= 1
        return ret

    def initIO(self, input_stream=sys.stdin, output_stream=sys.stdout):
        self.input_stream = input_stream
        self.output_stream = output_stream
        self.buffer = []
        self.buffer_index = 0

    def next(self):
        while not self.buffer:
            line = self.input_stream.readline().strip()
            if not line:
                break
            self.buffer.extend(line.split())
        if not self.buffer:
            return None
        return self.buffer.pop(0)

    def nextInt(self):
        return int(self.next())

    def nextLong(self):
        return int(self.next())

    def print(self, a):
        self.output_stream.write(str(a))

    def printsp(self, a):
        self.print(a)
        self.print(' ')

    def println(self):
        self.output_stream.write('\n')

    def flush(self):
        self.output_stream.flush()


if __name__ == "__main__":
    solver = Solver()
    solver.solve()
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/comment/direct/Java/C++/s316260310_commented.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Helper {
private:
    long long MOD;
    int MAXN;
public:
    Helper(long long mod, int maxn) : MOD(mod), MAXN(maxn) {}

    static const int BUFSIZE = 1 << 20;
    static char buf[BUFSIZE];
    static int index, total;

    static void initIO() {
        index = 0;
        total = 0;
    }

    static int scan() {
        if (index >= total) {
            index = 0;
            total = fread(buf, 1, BUFSIZE, stdin);
            if (total <= 0) return -1;
        }
        return buf[index++];
    }

    int nextInt() {
        int c;
        for (c = scan(); c <= 32; c = scan()) ;
        bool neg = false;
        if (c == '-' || c == '+') {
            neg = (c == '-');
            c = scan();
        }
        int val = 0;
        for (; c > 32; c = scan()) {
            val = val * 10 + (c - '0');
        }
        return neg ? -val : val;
    }

    long long nextLong() {
        int c;
        for (c = scan(); c <= 32; c = scan()) ;
        bool neg = false;
        if (c == '-' || c == '+') {
            neg = (c == '-');
            c = scan();
        }
        long long val = 0;
        for (; c > 32; c = scan()) {
            val = val * 10 + (c - '0');
        }
        return neg ? -val : val;
    }

    string next() {
        int c;
        for (c = scan(); c <= 32; c = scan()) ;
        string s;
        for (; c > 32; c = scan()) {
            s += (char)c;
        }
        return s;
    }

    vector<int> getIntArray(int size) {
        vector<int> ar(size);
        for (int i = 0; i < size; ++i) ar[i] = nextInt();
        return ar;
    }

    vector<long long> getLongArray(int size) {
        vector<long long> ar(size);
        for (int i = 0; i < size; ++i) ar[i] = nextLong();
        return ar;
    }

    void print(int a) { cout << a; }
    void print(long long a) { cout << a; }
    void printsp(int a) { cout << a << " "; }
    void println() { cout << endl; }
    void println(int a) { cout << a << endl; }
    void println(long long a) { cout << a << endl; }
    void flush() { cout.flush(); }

    long long gcd(long long a, long long b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
};

char Helper::buf[Helper::BUFSIZE];
int Helper::index = 0;
int Helper::total = 0;

class Solver {
private:
    Helper hp;
    const int MAXN = 1000006;
    const long long MOD = 1e9 + 7;

public:
    Solver() : hp(MOD, MAXN) {}

    void solve() {
        bool TESTCASES = false;
        int tc = TESTCASES ? hp.nextInt() : 1;
        for (int tce = 1; tce <= tc; ++tce) {
            solve(tce);
        }
        hp.flush();
    }

    void solve(int tc) {
        int N = hp.nextInt();
        long long K = hp.nextLong();
        vector<int> A = hp.getIntArray(N);

        vector<int> straight, cycle;
        vector<bool> vis(N, false);
        int curr = 0;
        while (!vis[curr]) {
            vis[curr] = true;
            curr = A[curr] - 1;
        }

        int st = 0;
        while (st != curr) {
            straight.push_back(st);
            st = A[st] - 1;
        }

        cycle.push_back(curr);
        curr = A[curr] - 1;
        while (curr != cycle[0]) {
            cycle.push_back(curr);
            curr = A[curr] - 1;
        }

        if (K < static_cast<long long>(straight.size())) {
            hp.println(straight[K] + 1);
        } else {
            long long rem = K - static_cast<long long>(straight.size());
            int idx = static_cast<int>(rem % cycle.size());
            hp.println(cycle[idx] + 1);
        }
    }
};

int main() {
    Solver solver;
    solver.solve();
    return 0;
}
// End of Code
